### PR TITLE
[RUNTIME] Runtime/student code upload setup

### PR DESCRIPTION
**Pull request by brandonxxlee**
When Dawn is about to upload student code, we make sure the studentCode process is not running.

### CODE REVIEW

File: runtime/testy/fake_dawn.py
Comment (nikitakit) [inline runtime/testy/fake_dawn.py:53]: What's up with this change?

It looks like it would make quite a big difference on the functionality of fake_dawn, but I fail to see how it relates to student code upload.

File: runtime/testy/runtime.py
Comment (nikitakit) [inline runtime/testy/runtime.py:94]: Has issue #95 been addressed yet?

File: runtime/testy/runtime.py
Comment (nikitakit) [inline runtime/testy/runtime.py]: Given that you added line 94 `controlState = "idle"`, I think you should also add it here and remove it from line 105.

File: runtime/testy/runtime.py
Comment (nikitakit) [inline runtime/testy/runtime.py]: It looks like you're moving all control state changes into the while loop, so I think you should move this one too (and maybe add an assert if the controlstate must always be idle at this point).

File: runtime/testy/runtime.py
Comment (nikitakit) [inline runtime/testy/runtime.py:159]: Note that this section of code is also touched by #113, which has basically passed code review. I think you guys should make an effort to get all changes merged into master asap, instead of dragging out the merge conflicts for a long time.

File: runtime/testy/stateManager.py
Comment (nikitakit) [inline runtime/testy/stateManager.py:147]: Would there ever be a situation where you write to ansible with `ANSIBLE_COMMANDS.STUDENT_UPLOAD`, but set the data bit to `False` rather than `True`?

**Review by nikitakit — CHANGES_REQUESTED** [review]
A few minor suggestions in comments, but mostly looks good.

File: runtime/testy/fake_dawn.py
Comment (bodiddles) [reply runtime/testy/fake_dawn.py:53]: The student upload message tells runtime to kill any running student code process as a safety measure so Dawn can scp over a new version of student code for us to run. We moved it out of the while loop so it only runs once as a command to kill any running student code which also mirrors actual dawn usage of this command.

File: runtime/testy/runtime.py
Comment (brandonxxlee) [reply runtime/testy/runtime.py:94]: We will address this in #153 to properly kill studentCode Process when swapping modes other than IDLE. (IDLE here should kill the student code.)

File: runtime/testy/stateManager.py
Comment (brandonxxlee) [reply runtime/testy/stateManager.py:147]: Theoretically we would send False if we are not ready for an upload, but as far as I can tell as long as we kill student code we should always be ready, so I guess this will just always be `True`.

File: runtime/testy/runtime.py
Comment (brandonxxlee) [reply runtime/testy/runtime.py]: Could you specify what you mean here? I'm not entirely sure what I'd be moving, or what the assert would be for.

Comment (nikitakit) [reply runtime/testy/runtime.py]: The line 94 you added looks half-redundant with this. Having both is confusing, but you do need to handle other cases.

Comment (brandonxxlee) [reply runtime/testy/runtime.py]: I have moved the `controlstate = "idle"` into both parts inside the while loop. Are there more things I should add/move?

File: runtime/testy/fake_dawn.py
Comment (nikitakit) [reply runtime/testy/fake_dawn.py:53]: sounds good

**Review by nikitakit — APPROVED** [review]
lgtm

**Review by yizow — APPROVED** [review]
LGTM

Make sure before you click the merge button to run all tests. Hate to break master, and it's happened a number of times already.

**Pull request merged**
